contracts/misc/P2PPairStaking.sol

```diff
@@ -0,0 +1,588 @@
+        address _bakc,
+        address _apeCoin,
+        address _cApe,
+        address _apeCoinStaking
+    ) {
+        bayc = _bayc;
+        mayc = _mayc;
+        bakc = _bakc;
+        apeCoin = _apeCoin;
+        cApe = _cApe;
+        apeCoinStaking = ApeCoinStaking(_apeCoinStaking);
+    }
+
+    function initialize() public initializer {
+        __Ownable_init();
+        __ReentrancyGuard_init();
+
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                EIP712_DOMAIN,
+                keccak256("ParaSpace"),
+                keccak256(bytes("1")),
+                block.chainid,
+                address(this)
+            )
+        );
+
+        //approve for apeCoinStaking
+        uint256 allowance = IERC20(apeCoin).allowance(
+            address(this),
+            address(apeCoinStaking)
+        );
+        if (allowance == 0) {
+            IERC20(apeCoin).safeApprove(
+                address(apeCoinStaking),
+                type(uint256).max
+            );
+        }
+
+        //approve for cApe
+        allowance = IERC20(apeCoin).allowance(address(this), cApe);
+        if (allowance == 0) {
+            IERC20(apeCoin).safeApprove(cApe, type(uint256).max);
+        }
+    }
+
+    function cancelListing(ListingOrder calldata listingOrder) external nonReentrant {
+        require(msg.sender == listingOrder.offerer, "not order offerer");
+        bytes32 orderHash = getListingOrderHash(listingOrder);
+        require(!isListingOrderCanceled[orderHash], "order already cancel");
+        isListingOrderCanceled[orderHash] = true;
+
+        emit OrderCancelled(orderHash, listingOrder.offerer);
+    }
+
+    function matchPairStakingList(
+        ListingOrder calldata apeOrder,
+        ListingOrder calldata apeCoinOrder
+    ) external nonReentrant returns (bytes32 orderHash) {
+        //1 validate all order
+        _validateApeOrder(apeOrder);
+        _validateApeCoinOrder(apeCoinOrder);
+
+        //2 check if orders can match
+        require(
+            apeOrder.stakingType == StakingType.BAYCStaking ||
+                apeOrder.stakingType == StakingType.MAYCStaking,
+            "invalid stake type"
+        );
+        require(
+            apeOrder.stakingType == apeCoinOrder.stakingType,
+            "orders type match failed"
+        );
+        require(
+            apeOrder.share + apeCoinOrder.share ==
+                PercentageMath.PERCENTAGE_FACTOR,
+            "orders share match failed"
+        );
+
+        //3 transfer token
+        IERC721(apeOrder.token).safeTransferFrom(
+            apeOrder.offerer,
+            address(this),
+            apeOrder.tokenId
+        );
+        uint256 apeAmount = getApeCoinStakingCap(apeOrder.stakingType);
+        IERC20(apeCoin).safeTransferFrom(
+            apeCoinOrder.offerer,
+            address(this),
+            apeAmount
+        );
+
+        //4 create match order
+        MatchedOrder memory matchedOrder;
```

Comment:
is this more gas effecient than using `var = StructStype({param1: value, param2: value});` ?

Reply:
I think it's almost same for memory struct. For storage struct, it'll consume more gas in my way. Anyway, it's better practice, i'll change it.